builtin/game/falling.lua

```diff
@@ -160,36 +160,86 @@ function nodeupdate_single(p, delay)
 					core.get_node_level(p_bottom) < core.get_node_max_level(p_bottom))) and
 				(not core.registered_nodes[n_bottom.name].walkable or
 					core.registered_nodes[n_bottom.name].buildable_to) then
-			if delay then
-				core.after(0.1, nodeupdate_single, p, false)
-			else
-				n.level = core.get_node_level(p)
-				core.remove_node(p)
-				spawn_falling_node(p, n)
-				nodeupdate(p)
-			end
+			n.level = core.get_node_level(p)
+			core.remove_node(p)
+			spawn_falling_node(p, n)
+			return true
 		end
 	end
 
 	if core.get_item_group(n.name, "attached_node") ~= 0 then
 		if not check_attached_node(p, n) then
 			drop_attached_node(p)
-			nodeupdate(p)
+			return true
 		end
 	end
+
+	return false
 end
 
+-- this table is specifically ordered.
+-- we don't walk diagonals, only our direct neighbors, and self
+-- down first as likely case, but always before self. same with sides.
+-- up must come last, so that things above self will also fall all at once
+local nodeupdate_neighbors = {
+	{x = 0, y = -1, z = 0},
+	{x = -1, y = 0, z = 0},
+	{x = 1, y = 0, z = 0},
+	{x = 0, y = 0, z = 1},
+	{x = 0, y = 0, z = -1},
+	{x = 0, y = 0, z = 0},
+	{x = 0, y = 1, z = 0},
+}
+
 function nodeupdate(p, delay)
 	-- Round p to prevent falling entities to get stuck
 	p = vector.round(p)
 
-	for x = -1, 1 do
-	for y = -1, 1 do
-	for z = -1, 1 do
-		local d = vector.new(x, y, z)
-		nodeupdate_single(vector.add(p, d), delay or not (x == 0 and y == 0 and z == 0))
-	end
-	end
+	-- we make a stack, and manually maintain size for performance
+	-- on the stack, we will maintain a table with pos and neigbor order
+	local s = {}
+	local n = 0
+	-- the neighbor order we will visit from our table
+	local v = 1
+
+	while true do
+		-- push current pos onto the stack
+		n = n + 1
+		s[n] = {p = p, v = v}
+		-- select next node from neighbor list
+		p = vector.add(p, nodeupdate_neighbors[v])
+		-- now we check out the node. If it is in need of an update,
+		-- it will let us know in the return value (true = updated)
+		if not nodeupdate_single(p) then
+			-- if we don't need to "recurse" (walk) to it then pop
+			-- our previous pos off the stack and continue from there
+			-- with the v value we were at when we last were at that
+			-- node
+			p = s[n].p
+			v = s[n].v
+			s[n] = nil
+			n = n - 1
+			-- aggressively prune tail, this could rewind many
+			-- last steps up in a row now
+			while v == 7 do
+				-- if there's nothing left on the stack, we're
+				-- done and can exit
+				if n == 0 then
+					return
+				end
+				p = s[n].p
+				v = s[n].v
+				s[n] = nil
+				n = n - 1
+			end
```

Comment:
You can merge the wind-up code above this loop with the loop by using a repeat-until loop.